greenlight/lib/api/prefs.py

```diff
@@ -2,109 +2,174 @@
+        :param interface: Interface to use for the complex value (default None)
+                          Possible values are: nsILocalFile, nsISupportsString,
+                          and nsIPrefLocalizedString
+
+        :returns: The value of the specified preference.
         """
+        assert pref_name
+
+        # Bug 1118825 - None is causing an exception to be raised
+        interface = interface or ''
+
         with self.marionette.using_context('chrome'):
             value = self.marionette.execute_script("""
-              let pref = arguments[0];
-              let prefBranch = Cc["@mozilla.org/preferences-service;1"]
-                              .getService(Ci.nsIPrefBranch);
-              let type = prefBranch.getPrefType(pref);
+              Cu.import("resource://gre/modules/Services.jsm");
+
+              let pref_name = arguments[0];
+              let default_branch = arguments[1];
+              let interface = arguments[2];
+
+              let prefBranch = undefined;
+              if (default_branch) {
+                prefBranch = Services.prefs.getDefaultBranch("");
+              }
+              else {
+                prefBranch = Services.prefs.QueryInterface(Ci.nsIPrefBranch);
+              }
+
+              // If an interface has been set, handle it differently
+              if (interface != '') {
+                return prefBranch.getComplexValue(pref_name,
+                                                  Ci[interface]).data;
+              }
+
+              let type = prefBranch.getPrefType(pref_name);
 
               switch (type) {
                 case prefBranch.PREF_STRING:
-                  return prefBranch.getCharPref(pref);
+                  return prefBranch.getCharPref(pref_name);
                 case prefBranch.PREF_BOOL:
-                  return prefBranch.getBoolPref(pref);
+                  return prefBranch.getBoolPref(pref_name);
                 case prefBranch.PREF_INT:
-                  return prefBranch.getIntPref(pref);
+                  return prefBranch.getIntPref(pref_name);
                 case prefBranch.PREF_INVALID:
                   return null;
               }
-            """, script_args=[pref])
+            """, script_args=[pref_name, default_branch, interface])
+
+        return value
 
-        return self._cast(value)
+    def set_pref(self, pref_name, value):
+        """Sets a preference to a specified value.
 
-    def set_pref(self, pref, value):
-        """
-        Sets the specified pref to value. Also archives the old value so that
-        the pref can be restored with `restore_pref`.
+        Also archives the old value so that the preference can be restored with
+        `restore_pref`.
 
-        :param pref: The preference to set.
+        :param pref_name: Name of the preference to set.
         :param value: The value to set the preference to.
         """
+        assert pref_name is not None
+        assert value is not None
```

Comment:
We do not check for the types of those values for consistency but to figure out which type a new pref has to be. We fail at the moment because so far we are trying to get the type via the stored pref, which in this case does not exist. So the assert as you propose here will only make sure that the right types are allowed for that method. As long as we do not use complex prefs this should be fine. Do you still want to have this?